handler/spa.go

```diff
@@ -19,48 +26,113 @@ var (
+	bootstrapOnce    sync.Once
+	config           *config.Spa
+	modifier         []hcl.Body
+	srvOptions       *server.Options
 }
 
-func NewSpa(bootstrapFile string, srvOpts *server.Options, modifier []hcl.Body) (*Spa, error) {
-	absPath, err := filepath.Abs(bootstrapFile)
-	if err != nil {
+func init() {
+	// related to bootstrap_data and ist a must due to XSS
+	ctyjson.EscapeHTML = true
+}
+
+func NewSpa(ctx *hcl.EvalContext, config *config.Spa, srvOpts *server.Options, modifier []hcl.Body) (*Spa, error) {
+	var err error
+	if config.BootstrapFile, err = filepath.Abs(config.BootstrapFile); err != nil {
 		return nil, err
 	}
-	return &Spa{
-		file:       absPath,
+
+	spa := &Spa{
+		config:     config,
 		modifier:   modifier,
 		srvOptions: srvOpts,
-	}, nil
+	}
+
+	if config.BootstrapData == nil {
+		return spa, nil
+	}
+
+	file, err := os.Open(config.BootstrapFile)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return nil, err
+	}
+	spa.bootstrapModTime = fileInfo.ModTime()
+
+	spa.replaceBootstrapData(ctx, file)
+
+	return spa, nil
 }
 
 func (s *Spa) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	file, err := os.Open(s.file)
-	if err != nil {
-		if _, ok := err.(*os.PathError); ok {
-			s.srvOptions.ServerErrTpl.WithError(errors.RouteNotFound).ServeHTTP(rw, req)
+	var content io.ReadSeeker
+	var modTime time.Time
+
+	if len(s.bootstrapContent) > 0 {
+		content = strings.NewReader(string(s.bootstrapContent))
+		modTime = s.bootstrapModTime
+	} else {
+		file, err := os.Open(s.config.BootstrapFile)
+		if err != nil {
+			if _, ok := err.(*os.PathError); ok {
+				s.srvOptions.ServerErrTpl.WithError(errors.RouteNotFound).ServeHTTP(rw, req)
+				return
+			}
+
+			s.srvOptions.ServerErrTpl.WithError(errors.Configuration).ServeHTTP(rw, req)
 			return
 		}
+		content = file
+		defer file.Close()
 
-		s.srvOptions.ServerErrTpl.WithError(errors.Configuration).ServeHTTP(rw, req)
-		return
-	}
-	defer file.Close()
+		fileInfo, err := file.Stat()
+		if err != nil || fileInfo.IsDir() {
+			s.srvOptions.ServerErrTpl.WithError(errors.Configuration).ServeHTTP(rw, req)
+			return
+		}
 
-	fileInfo, err := file.Stat()
-	if err != nil || fileInfo.IsDir() {
-		s.srvOptions.ServerErrTpl.WithError(errors.Configuration).ServeHTTP(rw, req)
-		return
+		modTime = fileInfo.ModTime()
 	}
 
+	evalContext := eval.ContextFromRequest(req)
+
 	if r, ok := rw.(*writer.Response); ok {
-		evalContext := eval.ContextFromRequest(req)
 		r.AddModifier(evalContext.HCLContext(), s.modifier...)
 	}
 
-	http.ServeContent(rw, req, s.file, fileInfo.ModTime(), file)
+	http.ServeContent(rw, req, s.config.BootstrapFile, modTime, content)
+}
+
+func (s *Spa) replaceBootstrapData(ctx *hcl.EvalContext, reader io.ReadCloser) {
+	if s.config.BootstrapData == nil {
+		return
+	}
+
+	b, err := io.ReadAll(reader)
+	if err != nil {
+		panic(err)
+	}
+
+	val, _ := s.config.BootstrapData.Value(ctx)
```

Comment:
1. handle `diags` caused by e.g. unknown variables
2. check `val` type: `val.Type().IsObjectType()` (must be object according to documentation)